youtubeApi/oauth2.go

```
@@ -0,0 +1,222 @@
+	}
+	fmt.Println(authURL)
+	return exchangeToken(config, code)
+}
+
+// getTokenFromWeb uses Config to request a Token.
+// It returns the retrieved Token.
+func getTokenFromWeb(config *oauth2.Config, authURL string) (*oauth2.Token, error) {
+	codeCh, err := startWebServer()
+	if err != nil {
+		fmt.Printf("Unable to start a web server.")
+		return nil, err
+	}
+
+	err = openURL(authURL)
+	if err != nil {
+		log.Fatalf("Unable to open authorization URL in web server: %v", err)
+	} else {
+		fmt.Println("Your browser has been opened to an authorization URL.", " This program will resume once authorization has been provided.\n")
+		fmt.Println(authURL)
+	}
+
+	// Wait for the web server to get the code.
+	code := <-codeCh
+	return exchangeToken(config, code)
+}
+
+// tokenCacheFile generates credential file path/filename.
+// It returns the generated credential path/filename.
+func tokenCacheFile() (string, error) {
+	usr, err := user.Current()
+	if err != nil {
+		return "", err
+	}
+	tokenCacheDir := filepath.Join(usr.HomeDir, ".credentials")
+	os.MkdirAll(tokenCacheDir, 0700)
+	return filepath.Join(tokenCacheDir,
+		url.QueryEscape("youtube-go.json")), err
+}
+
+// tokenFromFile retrieves a Token from a given file path.
+// It returns the retrieved Token and any read error encountered.
+func tokenFromFile(file string) (*oauth2.Token, error) {
+	f, err := os.Open(file)
+	if err != nil {
+		return nil, err
+	}
+	t := &oauth2.Token{}
+	err = json.NewDecoder(f).Decode(t)
+	defer f.Close()
+	return t, err
+}
+
+// saveToken uses a file path to create a file and store the
+// token in it.
+func saveToken(file string, token *oauth2.Token) {
+	fmt.Println("trying to save token")
+	fmt.Printf("Saving credential file to: %s\n", file)
+	f, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		log.Fatalf("Unable to cache oauth token: %v", err)
+	}
+	defer f.Close()
+	json.NewEncoder(f).Encode(token)
```

`Encode()` could return an error, but this error will be ignored.